x-pack/plugins/security/server/audit_trail/audit_trail_client.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { Subject } from 'rxjs';
+import { KibanaRequest, Auditor, AuditEvent, AuditEventDecorator } from 'src/core/server';
+
+import { SecurityPluginSetup } from '../';
+import { SpacesPluginSetup } from '../../../spaces/server';
+
+interface Deps {
+  getCurrentUser: SecurityPluginSetup['authc']['getCurrentUser'];
+  getSpaceId?: SpacesPluginSetup['spacesService']['getSpaceId'];
+}
+
+export class AuditTrailClient implements Auditor {
+  constructor(
+    private readonly request: KibanaRequest,
+    private readonly event$: Subject<AuditEvent>,
+    private readonly deps: Deps
+  ) {}
+
+  public add<Args>(decorateEvent: AuditEventDecorator<Args>, args: Args) {
+    const user = this.deps.getCurrentUser(this.request);
+    const spaceId = this.deps.getSpaceId?.(this.request);
+    const event = decorateEvent(
+      {
+        user: {
+          name: user?.username!,
+          roles: user?.roles,
```

Comment:
I think we'd need to add redactions feature if we want to add email field due to GDPR.
The other concern is that this would be unnecessary meta data for most events. I think username should be included for every event but the rest of the user object should be populated only for events that benefit from the extra information (e.g. login/logout)